### ISSUE TITLE
Wrong emulation of the procedural FOR loop with index

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11361]
The procedural `FOR` loop with index is currently emulated using `WHILE` in RDBMS that don't support `FOR`:

```java
begin(
    for_(i).in(1, 5).loop(
        insertInto(t).columns(a).values(i)
    ),
    for_(i).in(6, 10).loop(
        insertInto(t).columns(a).values(i)
    )
)
```

In HSQLDB, this produces (when placed in a procedure):

```sql
create procedure "proc"()
modifies sql data
"proc":
begin atomic
  declare "i" int default 1;
  while "i" <= 5 do
    insert into "t" ("a")
    values ("i");
    set "i" = ("i" + 1);
  end while;
  declare "i" int default 6;
  while "i" <= 10 do
    insert into "t" ("a")
    values ("i");
    set "i" = ("i" + 1);
  end while;
end;
```

But this is obviously wrong, `"i"` is already declared when it is set to `6` for the second loop. We must wrap the declaration in a new scope using `begin`:

This doesn't work for SQL Server, which doesn't know block scope: https://github.com/jOOQ/jOOQ/issues/8307

**Comment by lukaseder** [comment]
A related bug in HSQLDB: https://sourceforge.net/p/hsqldb/bugs/1610/

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.8 (#11362)

**Comment by lukaseder** [comment]
The correct SQL generation (in HSQLDB) is:

```sql
create procedure "proc"()
modifies sql data
"proc":
begin atomic
  begin atomic
    declare "i" int default 1;
    while "i" <= 5 do
      insert into "t" ("a")
      values ("i");
      set "i" = ("i" + 1);
    end while;
  end;
  begin atomic
    declare "i" int default 6;
    while "i" <= 10 do
      insert into "t" ("a")
      values ("i");
      set "i" = ("i" + 1);
    end while;
  end;
end;
```

**Comment by lukaseder** [comment]
The fix produces a regression in Firebird

This loop:

```java
for_(i).in(1, 10).by(2).loop(
    insertInto(t).columns(a).values(i)
)
```

Is emulated as such:

```sql
execute block as begin
    declare "i" integer default 1;
  begin
    while (:"i" <= 10) do begin
      insert into "t" ("a")
      values (:"i");
      :"i" = (:"i" + 2);
    end
  end
end
```

(The declarations are pulled up to the beginning in Firebird, which doesn't support variable declarations in nested scopes). If anything at all, the Firebird emulation should be this, instead:

```sql
execute block as 
  declare "i" integer;
begin
  :"i" = 1;
  while (:"i" <= 10) do begin
    insert into "t" ("a")
    values (:"i");
    :"i" = (:"i" + 2);
  end
end
```

I'll look into this a bit more throughly. When two subsequent loops re-use a loop variable, we must obviously declare it only once.

**Comment by lukaseder** [comment]
It's not a regression. We already have this problem today on https://www.jooq.org/translate/

![image](https://user-images.githubusercontent.com/734593/106792737-d6173d80-6656-11eb-953c-9d22aa2e7c64.png)

... whose snapshot version is already a bit older. Probably, we just skipped the test for now, because of a known issue (maybe also #8307). Will investigate this issue tomorrow.

**Comment by lukaseder** [comment]
The firebird issue has been extracted into a separate issue: https://github.com/jOOQ/jOOQ/issues/11366
